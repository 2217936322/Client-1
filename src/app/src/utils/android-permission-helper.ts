import { PermissionsAndroid } from 'react-native';

// for Android

/**
 * 请求安卓外部存储权限
 */
export const requestExternalStoragePermission = async (): Promise<boolean> => {
  try {
    const check = await PermissionsAndroid.check(
      PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE
    );
    if (check) {
      return true;
    }

    const granted = await PermissionsAndroid.request(
      PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
      {
        title: '请求读写权限',
        message: 'TRPG Engine 需要访问你的外部存储以保存图片',
        buttonPositive: '好的',
      }
    );

    if (granted === PermissionsAndroid.RESULTS.GRANTED) {
      return true;
    }

    return false;
  } catch (err) {
    console.error('Failed to request permission ', err);
    return false;
  }
};
