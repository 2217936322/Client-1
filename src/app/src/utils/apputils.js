import Toast from 'react-native-root-toast';

// 用于全局，一般情况下使用组件形式：<TToast />
export const toast = function(message, options) {
  let toast = Toast.show(
    message,
    Object.assign(
      {},
      {
        duration: Toast.durations.SHORT,
        position: Toast.positions.BOTTOM,
        shadow: false,
        animation: true,
        backgroundColor: 'rgba(0,0,0,0.8)',
        hideOnPress: true,
      },
      options
    )
  );

  return {
    element: toast,
    hide: () => Toast.hide(toast),
  };
};
