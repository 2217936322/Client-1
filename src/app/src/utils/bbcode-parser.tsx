import React, { ReactElement } from 'react';
import { Text } from 'react-native';
import { parse } from '@bbob/parser';
import ImageTag from './bbcode/ImageTag';
import { AstNode } from './bbcode/types';
import _has from 'lodash/has';
import _last from 'lodash/last';

const tagMap = {
  img: ImageTag,
};

/**
 * BBCode 解析器
 */
class BBCodeParser {
  options = {
    onlyAllowTags: ['img'],
    onError: (err) => {
      console.warn(err.message, err.lineNumber, err.columnNumber);
    },
  };

  parse(input: string): ReactElement[] {
    const ast = parse(input, this.options) as AstNode[];

    return ast
      .reduce<AstNode[]>((prev, curr) => {
        if (typeof curr === 'string' && typeof _last(prev) === 'string') {
          // 合并字符串, 使其渲染时能公用一个Text组件
          prev[prev.length - 1] += curr;
        } else {
          prev.push(curr);
        }

        return prev;
      }, [])
      .map<ReactElement>((node, index) => {
        if (typeof node === 'string') {
          return <Text key={index}>{node}</Text>;
        }

        if (typeof node === 'object' && _has(tagMap, node.tag)) {
          const Component = tagMap[node.tag];
          return <Component key={index} node={node} />;
        }

        return null;
      });
  }
}

const bbcodeParser = new BBCodeParser();

export default bbcodeParser;
