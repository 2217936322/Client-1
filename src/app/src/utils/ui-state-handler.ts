/**
 * 这是一个全局UI管理处理器
 * 用于生成将redux的状态转化为UI
 */

import TAlert from '../components/TApi/TAlert';
import { hideAlert, hideModal } from '@src/shared/redux/actions/ui';
import { Toast, Portal } from '@ant-design/react-native';
import TModal from '../components/TApi/TModal';
import { TRPGDispatch } from '@src/shared/redux/types/__all__';

type UIMap = { [name: string]: any };
interface FactoryOptions {
  onEnabled?: (currentUIState: UIMap, dispatch: TRPGDispatch) => void;
  onDisabled?: (currentUIState: UIMap, dispatch: TRPGDispatch) => void;
}

/**
 * 生成监听state ui状态变化的函数
 * @param variable 监听的变量
 * @param option 配置
 */
export const uiStateSwitchFactory = (
  variable: string,
  option: FactoryOptions
) => {
  return (prevUI: UIMap, currentUI: UIMap, dispatch: TRPGDispatch) => {
    if (prevUI.get(variable) === false && currentUI.get(variable) === true) {
      // UI打开
      option.onEnabled && option.onEnabled(currentUI, dispatch);
    }

    if (prevUI.get(variable) === true && currentUI.get(variable) === false) {
      // UI关闭
      option.onDisabled && option.onDisabled(currentUI, dispatch);
    }
  };
};

/**
 * Alert处理
 */
export const alertHandler = uiStateSwitchFactory('showAlert', {
  onEnabled: (currentUI, dispatch) => {
    TAlert.show(currentUI.get('showAlertInfo'), {
      onRequestClose: () => dispatch(hideAlert()),
    });
  },
  onDisabled: () => {
    TAlert.hide();
  },
});

/**
 * Toast 处理
 */
let currentToastKey: number = 0;
export const toastHandler = uiStateSwitchFactory('showToast', {
  onEnabled: (currentUI) => {
    if (currentToastKey > 0) {
      // 如果已存在Toast, 先把之前的移除
      Portal.remove(currentToastKey);
    }
    currentToastKey = Toast.show(currentUI.get('showToastText'), 0);
  },
  onDisabled: () => {
    Portal.remove(currentToastKey);
    currentToastKey = 0;
  },
});

/**
 * Modal 处理
 */
export const modelHandler = uiStateSwitchFactory('showModal', {
  onEnabled: (currentUI, dispatch) => {
    const body = currentUI.get('showModalBody');
    if (body) {
      TModal.show(body, { onRequestClose: () => dispatch(hideModal()) });
    }
  },
  onDisabled: () => {
    TModal.hide();
  },
});

export const uiHandlerCollection = [alertHandler, toastHandler, modelHandler];
