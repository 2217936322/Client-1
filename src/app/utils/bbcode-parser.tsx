import React, { ReactElement } from 'react';
import { Text } from 'react-native';
import { parse } from '@bbob/parser';
import ImageTag from './bbcode/ImageTag';
import { AstNode } from './bbcode/types';
import _has from 'lodash/has';

const tagMap = {
  img: ImageTag,
};

/**
 * BBCode 解析器
 */
class BBCodeParser {
  options = {
    onlyAllowTags: ['img'],
    onError: (err) => {
      console.warn(err.message, err.lineNumber, err.columnNumber);
    },
  };

  parse(input: string): ReactElement[] {
    const ast = parse(input, this.options) as AstNode[];

    return ast.map<ReactElement>((node, index) => {
      if (typeof node === 'string') {
        return <Text key={index}>{node}</Text>;
      }

      if (typeof node === 'object' && _has(tagMap, node.tag)) {
        const Component = tagMap[node.tag];
        return <Component key={index} node={node} />;
      }

      return null;
    });
  }
}

const bbcodeParser = new BBCodeParser();

export default bbcodeParser;
