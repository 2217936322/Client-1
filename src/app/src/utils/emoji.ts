import emoji from 'node-emoji';
import { getCodeList } from '../../../shared/utils/emoji';
import _forEach from 'lodash/forEach';

const codeList = getCodeList();

export const emojiMap = _forEach(getCodeList(), (list, catalog, map) => {
  map[catalog] = list
    .filter((name) => emoji.hasEmoji(name))
    .map((name) => ({
      name,
      code: emoji.get(name),
    }));
});

export const emojiCatalog = [
  'people',
  'nature',
  'objects',
  'places',
  'symbols',
];

/**
 * 将带有emoji代码的文本转化为标准文本
 * @param {string} code 带有emoji代码的文本
 */
export function unemojify(code) {
  return emoji.unemojify(code);
}

/**
 * 将标准字符串转化为表情
 * @param {string} plainText 无格式字符串
 */
export function emojify(plainText) {
  return emoji.emojify(plainText);
}
