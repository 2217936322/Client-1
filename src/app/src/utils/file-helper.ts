import { Platform } from 'react-native';
import CameraRoll from '@react-native-community/cameraroll';
import { downloadFile, DocumentDirectoryPath } from 'react-native-fs';
import _last from 'lodash/last';
import { requestExternalStoragePermission } from './android-permission-helper';
import { logger } from '@shared/api/logger';

/**
 * 保存图片到本地相册
 * @param imageUrl 图片远程地址
 */
export async function saveImageToLocal(imageUrl: string): Promise<string> {
  logger.info(`save image to local in ${Platform.OS}: ${imageUrl}`);
  if (Platform.OS === 'ios') {
    // TODO: 待实现。暂时不支持ios
    // https://reactnative.cn/docs/cameraroll.html
    logger.warn('Not Support IOS yet');
    throw new Error('Not Support IOS yet');
  } else if (Platform.OS === 'android') {
    const pathname = _last(imageUrl.split('/'));
    const downloadDest = `${DocumentDirectoryPath}/${pathname}`;

    const res = await downloadFile({
      fromUrl: imageUrl,
      toFile: downloadDest,
      progress: (res) => {
        logger.info(res.bytesWritten, res.contentLength);
      },
    }).promise;

    logger.info('res', res);

    if (res && res.statusCode === 200) {
      const grant = await requestExternalStoragePermission();
      if (!grant) {
        throw new Error('Require External Storage Permission');
      }
      logger.info('downloadDest', downloadDest);
      const uri = await CameraRoll.saveToCameraRoll('file://' + downloadDest);
      logger.info('save to local', uri);
      return uri;
    } else {
      throw new Error('Download Image Failed');
    }
  } else {
    logger.warn('Not Support OS Type');
    throw new Error('Not Support OS Type');
  }
}
