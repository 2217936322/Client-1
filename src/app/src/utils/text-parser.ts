import { emojify } from './emoji';
import bbcodeParser from './bbcode-parser';
import urlRegex from 'url-regex';

/**
 * 解析文本
 * @param {string} plainText 服务端标准文本
 */
export function parse(plainText: string) {
  return bbcodeParser.parse(preProcessText(plainText));
}

/**
 * 客户端预处理文本
 * @param plainText 服务端文本
 */
export function preProcessText(plainText: string): string {
  let text = emojify(plainText); // 将:<emojicode>: 转化为 unicode字符
  text = text.replace(
    new RegExp(urlRegex({ exact: false, strict: true }), 'g'),
    '[url]$&[/url]'
  ); // 将聊天记录中的url提取成bbcode
  return text;
}
