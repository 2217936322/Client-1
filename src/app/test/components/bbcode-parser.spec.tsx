import React, { Fragment } from 'react';
import renderer from 'react-test-renderer';
import { preProcessText, parse } from '@src/app/src/utils/text-parser';

describe('bbcode-parser', () => {
  describe('preprocess', () => {
    it('simple url parse', () => {
      const text = preProcessText('http://baidu.com');
      expect(text).toBe('[url]http://baidu.com[/url]');
    });

    it('mix text and url parse', () => {
      const text = preProcessText('open:http://baidu.com');
      expect(text).toBe('open:[url]http://baidu.com[/url]');
    });

    it('mix text and more url parse', () => {
      const text = preProcessText(
        'open:http://baidu.com and http://google.com'
      );
      expect(text).toBe(
        'open:[url]http://baidu.com[/url] and [url]http://google.com[/url]'
      );
    });
  });

  // 文本解析成组件测试
  describe('text parse component', () => {
    const cases = [
      {
        title: 'plain text',
        text: 'TestString',
      },
      {
        title: 'plain text with space',
        text: 'Test String',
      },
      {
        title: 'pure image',
        text: '[img]http://exampleimg.com[/img]',
      },
      {
        title: 'mix image and text',
        text: '测试文字1[img]http://exampleimg.com[/img]测试文字2',
      },
      {
        title: 'simple url',
        text: 'http://baidu.com',
      },
      {
        title: 'mix url and text',
        text: '百度:http://baidu.com',
      },
      {
        title: 'mix more url and text',
        text: '百度:http://baidu.com 谷歌:http://google.com',
      },
      {
        title: 'mix url and text with no separation',
        text: '百度http://baidu.com',
      },
      {
        title: 'mix more url and text with no separation',
        text: '百度http://baidu.com谷歌http://google.com',
      },
      {
        title: 'mix url and image and text',
        text: '百度:[img]http://exampleimg.com[/img] http://baidu.com',
      },
    ];

    for (const item of cases) {
      it(item.title, () => {
        const Component = parse(item.text);
        const result = renderer
          .create(<Fragment>{Component}</Fragment>)
          .toJSON();
        expect(result).toMatchSnapshot();
      });
    }
  });
});
