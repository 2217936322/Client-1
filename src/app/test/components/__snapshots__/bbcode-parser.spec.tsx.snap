// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bbcode-parser text parse component mix image and text 1`] = `
Array [
  <Text>
    测试文字1
  </Text>,
  <View>
    <Modal
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    >
      <ImageViewer
        enableSwipeDown={true}
        imageUrls={
          Array [
            Object {
              "url": "http://exampleimg.com",
            },
          ]
        }
        index={0}
        menuContext={
          Object {
            "cancel": "取消",
            "saveToLocal": "保存到相册",
          }
        }
        onClick={[Function]}
        onSave={[Function]}
        onSwipeDown={[Function]}
        saveToLocalByLongPress={true}
      />
    </Modal>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 120,
              "width": 180,
            },
          ]
        }
      >
        <FastImageView
          onFastImageLoad={[Function]}
          resizeMode="contain"
          source={
            Object {
              "cache": "immutable",
              "priority": "normal",
              "uri": "http://exampleimg.com",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
  </View>,
  <Text>
    测试文字2
  </Text>,
]
`;

exports[`bbcode-parser text parse component mix more url and text 1`] = `
Array [
  <Text>
    百度:
  </Text>,
  <Link>
    http://baidu.com
  </Link>,
  <Text>
     谷歌:
  </Text>,
  <Link>
    http://google.com
  </Link>,
]
`;

exports[`bbcode-parser text parse component mix more url and text with no separation 1`] = `
Array [
  <Text>
    百度
  </Text>,
  <Link>
    http://baidu.com谷歌http
  </Link>,
  <Text>
    :
  </Text>,
  <Link>
    //google.com
  </Link>,
]
`;

exports[`bbcode-parser text parse component mix url and image and text 1`] = `
Array [
  <Text>
    百度:
  </Text>,
  <View>
    <Modal
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    >
      <ImageViewer
        enableSwipeDown={true}
        imageUrls={
          Array [
            Object {
              "url": "http://exampleimg.com",
            },
          ]
        }
        index={0}
        menuContext={
          Object {
            "cancel": "取消",
            "saveToLocal": "保存到相册",
          }
        }
        onClick={[Function]}
        onSave={[Function]}
        onSwipeDown={[Function]}
        saveToLocalByLongPress={true}
      />
    </Modal>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 120,
              "width": 180,
            },
          ]
        }
      >
        <FastImageView
          onFastImageLoad={[Function]}
          resizeMode="contain"
          source={
            Object {
              "cache": "immutable",
              "priority": "normal",
              "uri": "http://exampleimg.com",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
  </View>,
  <Text>
     
  </Text>,
  <Link>
    http://baidu.com
  </Link>,
]
`;

exports[`bbcode-parser text parse component mix url and text 1`] = `
Array [
  <Text>
    百度:
  </Text>,
  <Link>
    http://baidu.com
  </Link>,
]
`;

exports[`bbcode-parser text parse component mix url and text with no separation 1`] = `
Array [
  <Text>
    百度
  </Text>,
  <Link>
    http://baidu.com
  </Link>,
]
`;

exports[`bbcode-parser text parse component plain text 1`] = `
<Text>
  TestString
</Text>
`;

exports[`bbcode-parser text parse component plain text with space 1`] = `
<Text>
  Test String
</Text>
`;

exports[`bbcode-parser text parse component pure image 1`] = `
<View>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <ImageViewer
      enableSwipeDown={true}
      imageUrls={
        Array [
          Object {
            "url": "http://exampleimg.com",
          },
        ]
      }
      index={0}
      menuContext={
        Object {
          "cancel": "取消",
          "saveToLocal": "保存到相册",
        }
      }
      onClick={[Function]}
      onSave={[Function]}
      onSwipeDown={[Function]}
      saveToLocalByLongPress={true}
    />
  </Modal>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "height": 120,
            "width": 180,
          },
        ]
      }
    >
      <FastImageView
        onFastImageLoad={[Function]}
        resizeMode="contain"
        source={
          Object {
            "cache": "immutable",
            "priority": "normal",
            "uri": "http://exampleimg.com",
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`bbcode-parser text parse component simple url 1`] = `
<Link>
  http://baidu.com
</Link>
`;
